
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../4_optimising_snakemake/">
      
      
        <link rel="next" href="../6_hpc_snakemake/">
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.34">
    
    
      
        <title>Being reproducible with Snakemake - Reproducible and Scalable Research with Snakemake</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.35f28582.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  <!-- Matomo -->
<script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
      var u="https://matomo.sib.swiss/";
      _paq.push(['setTrackerUrl', u+'matomo.php']);
      _paq.push(['setSiteId', '220']);
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
    })();
  </script>
  <!-- End Matomo Code -->
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
  <a href="../../.."> 
    <strong>Click here to go to latest.</strong>
  </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Reproducible and Scalable Research with Snakemake" class="md-header__button md-logo" aria-label="Reproducible and Scalable Research with Snakemake" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Reproducible and Scalable Research with Snakemake
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Being reproducible with Snakemake
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/containers-snakemake-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/containers-snakemake-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Reproducible and Scalable Research with Snakemake" class="md-nav__button md-logo" aria-label="Reproducible and Scalable Research with Snakemake" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Reproducible and Scalable Research with Snakemake
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/containers-snakemake-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/containers-snakemake-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Course material
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Course material
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    General guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2_introduction_snakemake/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Snakemake
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3_generalising_snakemake/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Making a more general-purpose Snakemake workflow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4_optimising_snakemake/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Decorating and optimising a Snakemake workflow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Being reproducible with Snakemake
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Being reproducible with Snakemake
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-from-previous-session" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow from previous session
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#creating-a-rule-to-gather-read-counts" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a rule to gather read counts
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a rule to gather read counts">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-the-general-rule-structure" class="md-nav__link">
    <span class="md-ellipsis">
      Building the general rule structure
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gathering-the-input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Gathering the input files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-a-python-script-in-snakemake" class="md-nav__link">
    <span class="md-ellipsis">
      Using a Python script in Snakemake
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using a Python script in Snakemake">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Getting the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deciding-how-to-run-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Deciding how to run the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#providing-a-rule-specific-conda-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Providing a rule-specific conda environment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#adapting-the-snakefile-and-running-the-rule" class="md-nav__link">
    <span class="md-ellipsis">
      Adapting the Snakefile and running the rule
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#creating-a-rule-to-detect-differentially-expressed-genes-deg" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a rule to detect Differentially Expressed Genes (DEG)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a rule to detect Differentially Expressed Genes (DEG)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-the-general-rule-structure-again" class="md-nav__link">
    <span class="md-ellipsis">
      Building the general rule structure (again)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-an-r-script-in-snakemake" class="md-nav__link">
    <span class="md-ellipsis">
      Using an R script in Snakemake
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using an R script in Snakemake">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#downloading-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Downloading the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Running the script
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Running the workflow
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6_hpc_snakemake/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Snakemake in an HPC environment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7_debugging_snakemake/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional concepts and workflow design/debugging
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-from-previous-session" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow from previous session
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#creating-a-rule-to-gather-read-counts" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a rule to gather read counts
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a rule to gather read counts">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-the-general-rule-structure" class="md-nav__link">
    <span class="md-ellipsis">
      Building the general rule structure
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gathering-the-input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Gathering the input files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-a-python-script-in-snakemake" class="md-nav__link">
    <span class="md-ellipsis">
      Using a Python script in Snakemake
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using a Python script in Snakemake">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Getting the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deciding-how-to-run-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Deciding how to run the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#providing-a-rule-specific-conda-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Providing a rule-specific conda environment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#adapting-the-snakefile-and-running-the-rule" class="md-nav__link">
    <span class="md-ellipsis">
      Adapting the Snakefile and running the rule
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#creating-a-rule-to-detect-differentially-expressed-genes-deg" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a rule to detect Differentially Expressed Genes (DEG)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a rule to detect Differentially Expressed Genes (DEG)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-the-general-rule-structure-again" class="md-nav__link">
    <span class="md-ellipsis">
      Building the general rule structure (again)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-an-r-script-in-snakemake" class="md-nav__link">
    <span class="md-ellipsis">
      Using an R script in Snakemake
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using an R script in Snakemake">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#downloading-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Downloading the script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Running the script
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Running the workflow
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Being reproducible with Snakemake</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Use an input function to work with an unknown number of inputs</li>
<li>Run scripts from other languages (Python and R)</li>
<li>Deploy a rule-specific conda environment</li>
<li>Deploy a rule-specific Docker/Apptainer container</li>
</ul>
<h2 id="material">Material</h2>
<p><a class="md-button" href="../../assets/pdf/5_reproducibility_snakemake.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M0 64C0 28.7 28.7 0 64 0h160v128c0 17.7 14.3 32 32 32h128v144H176c-35.3 0-64 28.7-64 64v144H64c-35.3 0-64-28.7-64-64zm384 64H256V0zM176 352h32c30.9 0 56 25.1 56 56s-25.1 56-56 56h-16v32c0 8.8-7.2 16-16 16s-16-7.2-16-16V368c0-8.8 7.2-16 16-16m32 80c13.3 0 24-10.7 24-24s-10.7-24-24-24h-16v48zm96-80h32c26.5 0 48 21.5 48 48v64c0 26.5-21.5 48-48 48h-32c-8.8 0-16-7.2-16-16V368c0-8.8 7.2-16 16-16m32 128c8.8 0 16-7.2 16-16v-64c0-8.8-7.2-16-16-16h-16v96zm80-112c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v32h32c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V368"/></svg></span> Download the presentation</a></p>
<h2 id="workflow-from-previous-session">Workflow from previous session</h2>
<p>If you didn&rsquo;t finish the previous part or didn&rsquo;t do the optional exercises, you can restart from fully commented snakefiles, with a supplementary .fastq files quality check rule and multithreading implemented in all rules. You can download all the files (workflow and config) <a href="https://github.com/sib-swiss/containers-snakemake-training/tree/main/docs/solutions/session3">here</a> or copy them after cloning the course repository locally:
<div class="highlight"><pre><span></span><code>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/sib-swiss/containers-snakemake-training.git<span class="w">  </span><span class="c1"># Clone repository</span>
<span class="nb">cd</span><span class="w"> </span>containers-snakemake-training/<span class="w">  </span><span class="c1"># Go in repository</span>
cp<span class="w"> </span>-r<span class="w"> </span>docs/solutions/session3<span class="w"> </span>destination_path<span class="w">  </span><span class="c1"># Copy folder where needed; adapt destination_path to required path</span>
</code></pre></div></p>
<h2 id="exercises">Exercises</h2>
<p>In this series of exercises, you will create the last two rules of the workflow. Each rule will execute a script (one in Python and one in R; don&rsquo;t worry, this is not a programming course, so we wrote the scripts for you!), and both rules will have dedicated environments that you will need to take into account in the snakefiles.</p>
<details class="tip">
<summary>Development and back-up</summary>
<p>During this session as well, you will modify your Snakefile quite heavily, so it may be a good idea to make back-ups from time to time (with <code>cp</code> or a simple copy/paste) or use a versioning system. As a general rule, if you have a doubt on the code you are developing, do not hesitate to make a back-up beforehand.</p>
</details>
<h3 id="creating-a-rule-to-gather-read-counts">Creating a rule to gather read counts</h3>
<p>To perform a Differential Expression Analysis (DEA), it is easier to have a single file gathering all the read counts from the different samples. The next rule that you will create will both find the required files and merge them, thanks to an input function and a Python script.</p>
<h4 id="building-the-general-rule-structure">Building the general rule structure</h4>
<p>We already wrote the common elements of the rule so that you can focus on the most interesting parts (the missing <code>input</code> and the missing elements at the end):
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">rule</span> <span class="n">count_table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule merges gene count tables of an assembly into one table.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="s1">&#39;results/total_count_table.tsv&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/total_count_table.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
</code></pre></div></td></tr></table></div></p>
<p>Here, there is no <code>wildcards</code> in the rule: only one file will be created, and its name will not change depending on sample names because we use <strong>all</strong> the samples to create it.</p>
<details class="tip">
<summary>Explicit is better than implicit</summary>
<p>Even if a software cannot multithread, it is useful to add <code>threads: 1</code> to keep the syntax consistent between rules and clearly state that the software works with a single thread.</p>
</details>
<p><strong>Exercise:</strong> Given that this rule and the next one will be quite different from the previous ones, it is a good idea to implement them in a new snakefile. Create a new file <code>workflow/rules/analyses.smk</code> and copy the previous rule structure in it. Do you need to change anything else to use this rule in your workflow?</p>
<details class="success">
<summary>Answer</summary>
<p>To actually use this rule, Snakemake needs to be aware that it exists: this is done with the <code>include</code> statement. We need to add the following lines to the main Snakefile (<code>workflow/Snakefile</code>):
<div class="highlight"><pre><span></span><code><span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/analyses.smk&#39;</span>
</code></pre></div></p>
<p>The Snakefile will look like this:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">Main Snakefile of RNAseq analysis workflow. This workflow can clean and</span>
<span class="sd">map reads, and perform Differential Expression Analyses.</span>
<span class="sd">&#39;&#39;&#39;</span>

<span class="c1"># Config file path</span>
<span class="n">configfile</span><span class="p">:</span> <span class="s1">&#39;config/config.yaml&#39;</span>

<span class="c1"># Rules to execute workflow</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/read_mapping.smk&#39;</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/analyses.smk&#39;</span>

<span class="c1"># Master rule used to launch workflow</span>
<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Dummy rule to automatically generate required outputs.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">reads_quantification_genes</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">gene_level</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">])</span>
</code></pre></div></td></tr></table></div></p>
</details>
<p>Let&rsquo;s start filling those missing elements!</p>
<h4 id="gathering-the-input-files">Gathering the input files</h4>
<p>This task is quite complex: we need a way to identify all the rule inputs and gather them in a Python list. Here, there are only six samples, so in theory, you could list them directly&hellip; However, it isn&rsquo;t good practice and it quickly becomes un-manageable when the number of sample increases. Fortunately, there is a much more elegant and convenient solution: an input function, which provides the added benefit of scaling up very well.</p>
<p>We wrote one for you:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Input function used in rule count_table</span>
<span class="k">def</span> <span class="nf">get_gene_counts</span><span class="p">(</span><span class="n">wildcards</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function lists count tables from every sample in the config file</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">return</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;results/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_genes_read_quantification.tsv&quot;</span>
            <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]]</span>
</code></pre></div></td></tr></table></div></p>
<details class="warning">
<summary>Snakemake wildcards vs Python f-strings</summary>
<p>This input function is pure Python code: in the return statement, <code>{sample}</code> isn&rsquo;t a wildcard, it is an f-string variable! This shows that you can natively use basic Python elements in a workflow: Snakemake will still be able understand them. This is because Snakemake was built on top of Python.</p>
</details>
<p>This function will loop over the list of samples in the config file and replace <code>{sample}</code> with the current sample name of the iteration to create a string which is the output path from the rule <code>reads_quantification_genes</code> of said sample. Then, it will aggregate all the paths in a list and return this list.</p>
<details class="info">
<summary>More on input functions</summary>
<ul>
<li>Input functions take the <code>wildcards</code> global object as <strong>single argument</strong></li>
<li>You can access wildcard values inside an input function with the syntax <code>{wildcards.wildcards_name}</code></li>
<li>Input functions can return lists of files, which will be automatically handled like multiple inputs by Snakemake<ul>
<li>Input functions can also return a dictionary; in this case, the function should be called with the <code>unpack()</code> function:
<div class="highlight"><pre><span></span><code><span class="nb">input</span><span class="p">:</span> <span class="n">unpack</span><span class="p">(</span><span class="o">&lt;</span><span class="n">function_name</span><span class="o">&gt;</span><span class="p">)</span>
</code></pre></div>
The dictionary keys will be used as input names and the dictionary values will be used as input values, providing a list of named inputs</li>
</ul>
</li>
</ul>
</details>
<details class="warning">
<summary>Input functions and output directory</summary>
<p>Input functions are evaluated <strong>before</strong> the workflow is executed, so they cannot be used to list the content of an output directory, since it does not exist before the workflow is executed. Instead, you can use a <a href="https://snakemake.readthedocs.io/en/v8.20.5/snakefiles/rules.html#data-dependent-conditional-execution">checkpoint</a> to trigger a re-evaluation of the DAG.</p>
</details>
<p><strong>Exercise:</strong> Insert the function <code>get_gene_counts()</code> in <code>workflow/rules/analysis.smk</code> and adapt the input value of <code>count_table</code> accordingly. Do you need to insert the function in a specific location?</p>
<details class="success">
<summary>Answer</summary>
<p>The first step is to add the input function to the file. However, it needs to appear <strong>before the rule <code>count_table</code></strong>, otherwise we will see the error <code>name 'get_gene_counts' is not defined</code>. In other words, the function needs to be defined <strong>before</strong> Snakemake looks for it when it parses the input. Then, we need to set the function name as the rule input value.</p>
<p>The modular snakefile <code>workflow/rules/analysis.smk</code> will resemble this:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Input function used in rule count_table</span>
<span class="k">def</span> <span class="nf">get_gene_counts</span><span class="p">(</span><span class="n">wildcards</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function lists count tables from every sample in the config file</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">return</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;results/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_genes_read_quantification.tsv&quot;</span>
            <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]]</span>

<span class="n">rule</span> <span class="n">count_table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule merges gene count tables of an assembly into one table.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
<span class="hll">        <span class="n">get_gene_counts</span>  <span class="c1"># Add input function to rule</span>
</span>    <span class="n">output</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="s1">&#39;results/total_count_table.tsv&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/total_count_table.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
</code></pre></div></td></tr></table></div>
<strong>You don&rsquo;t need to use parentheses or specify any argument when you call an input function in the <code>input</code> directive.</strong> Doing so would actually change Snakemake behaviour!</p>
</details>
<p>Now that the rule inputs are defined, we need to set-up the script to process them.</p>
<h4 id="using-a-python-script-in-snakemake">Using a Python script in Snakemake</h4>
<h5 id="getting-the-script">Getting the script</h5>
<p>The counts will be concatenated thanks to a script called <code>count_table.py</code>. It was written in <a href="https://www.python.org/">Python</a>, takes a list of files as input, and produces one output, a tab-separated table containing read counts of the different samples for each gene. You can download it with:
<div class="highlight"><pre><span></span><code>wget<span class="w"> </span>https://raw.githubusercontent.com/sib-swiss/containers-snakemake-training/main/docs/solutions/session4/workflow/scripts/count_table.py
</code></pre></div></p>
<p>Or you can copy it from here:</p>
<details class="tip">
<summary>Click here to see a nice Python script!</summary>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">count_table.py</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">Merge gene counts from all samples of an assembly into a single table.</span>
<span class="sd">&#39;&#39;&#39;</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="hll"><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>  <span class="c1"># Non built-in package</span>
</span><span class="kn">import</span> <span class="nn">sys</span>


<span class="c1"># Constants</span>
<span class="n">FIELDS</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Geneid&#39;</span><span class="p">,</span> <span class="s1">&#39;Reads_quant&#39;</span><span class="p">]</span>
<span class="n">STR_TO_REMOVE</span> <span class="o">=</span> <span class="s1">&#39;_genes_read_quantification.tsv&#39;</span>


<span class="c1"># Functions</span>
<span class="k">def</span> <span class="nf">import_clean</span><span class="p">(</span><span class="n">table</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Importing and cleaning quantification data from &lt;</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s1">&gt;&#39;</span><span class="p">)</span>
    <span class="n">reads</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>
    <span class="n">reads</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">reads</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span> <span class="s1">&#39;Reads_quant&#39;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Sorting &lt;gene&gt; table by Chromosome then Start position&#39;</span><span class="p">)</span>
    <span class="c1"># New columns are simpler and will be used to properly reorder the table</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Creating temporary columns&#39;</span><span class="p">)</span>
    <span class="c1"># Get unique Chr ID using a set</span>
    <span class="n">reads</span><span class="p">[</span><span class="s1">&#39;Chr_new&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="s1">&#39;Chr&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))))</span>
    <span class="c1"># Select start of the first exon</span>
    <span class="n">reads</span><span class="p">[</span><span class="s1">&#39;Start_new&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Sorting table&#39;</span><span class="p">)</span>
    <span class="n">reads</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;Chr_new&#39;</span><span class="p">,</span> <span class="s1">&#39;Start_new&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">],</span>
                      <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Removing temporary columns&#39;</span><span class="p">)</span>
    <span class="n">reads</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;Chr_new&#39;</span><span class="p">,</span> <span class="s1">&#39;Start_new&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">final_table</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="n">FIELDS</span><span class="p">]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Geneid&#39;</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">final_table</span>


<span class="c1"># Main code execution</span>
<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">logfile</span><span class="p">:</span>

        <span class="c1"># Redirect everything from the script to Snakemake log</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span> <span class="o">=</span> <span class="n">logfile</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Getting data from snakemake&#39;</span><span class="p">)</span>
        <span class="n">list_of_files</span> <span class="o">=</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">input</span>
        <span class="n">count_table</span> <span class="o">=</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">table</span>

        <span class="n">output_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">count_table</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Initialising global table with &lt;</span><span class="si">{</span><span class="n">list_of_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&gt;&#39;</span><span class="p">)</span>
        <span class="n">total_table</span> <span class="o">=</span> <span class="n">import_clean</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">list_of_files</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Adding data from &lt;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s1">&gt;&#39;</span><span class="p">)</span>
            <span class="n">tmp_table</span> <span class="o">=</span> <span class="n">import_clean</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="n">total_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">total_table</span><span class="p">,</span> <span class="n">tmp_table</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Renaming columns&#39;</span><span class="p">)</span>
        <span class="n">column_titles</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">STR_TO_REMOVE</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
                         <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">list_of_files</span><span class="p">]</span>
        <span class="n">total_table</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">column_titles</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Saving final table in &lt;</span><span class="si">{</span><span class="n">count_table</span><span class="si">}</span><span class="s1">&gt;&#39;</span><span class="p">)</span>
        <span class="n">total_table</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">count_table</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
</details>
<p><strong>Exercise:</strong> Get the script with your favourite method and place it the proper folder according to the <a href="https://snakemake.readthedocs.io/en/v8.20.5/snakefiles/deployment.html">official documentation</a>. Where should you store it?</p>
<details class="success">
<summary>Answer</summary>
<p>Scripts should be gathered in their own dedicated folder: <code>workflow/scripts</code>.</p>
</details>
<h5 id="deciding-how-to-run-the-script">Deciding how to run the script</h5>
<p>If you remember the presentation, there are two directives that you can use to run external scripts in Snakemake: <code>run</code> and <code>script</code>. While both allow to run Python code, they are not equivalent, so there is a choice to make!</p>
<p><strong>Exercise:</strong> Check out the script content. Depending on what you find, choose a directive and implement it in place of the last two missing elements (<code>?</code>) of rule <code>count_table</code>.</p>
<details class="tip">
<summary>Script path is relative&hellip;</summary>
<p>&hellip; to the Snakefile calling it. If you followed the recommended workflow structure, modular snakefiles are placed in a <code>rules</code> subfolder (like <code>rules/analysis.smk</code>) and scripts are placed in a <code>scripts</code> subfolder (like <code>scripts/count_table.py</code>). You need to find a path between those two subfolders.</p>
</details>
<details class="success">
<summary>Answer</summary>
<p>There are two things to check before deciding which directive to use:</p>
<ol>
<li>
<p>The script length:</p>
<p>If we open the script in a text editor or run <code>wc -l workflow/scripts/count_table.py</code>, we see that it is 67 lines long. It is also quite complex, with function definitions, loops&hellip; This favours the <code>script</code> directive, as it&rsquo;s better to use <code>run</code> with short and simple code.</p>
</li>
<li>
<p>The use of external packages (packages that are not included in a default Python installation):</p>
<p>Another way to put this is: does the script need a special environment to work? If so, then we <strong>have to</strong> use the <code>script</code> directive, as it is the only one to accommodate for conda environments or containers. This means that this criteria takes precedence over the previous one: if we need to run a short script within a dedicated environment, <code>script</code> is the only way to do it.</p>
<p>Here, there are several <code>import</code> statements at the top of the script, including <code>import pandas as pd  # Non built-in package</code>. <a href="https://pandas.pydata.org/docs/index.html"><code>pandas</code></a> is a great package, but it is not part of the built-in packages shipped with Python. This means that the script needs a dedicated environment to run and confirm that we need the <code>script</code> directive.</p>
</li>
</ol>
<p>With this in mind, the rule will be:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">rule</span> <span class="n">count_table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule merges gene count tables of an assembly into one table.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">get_gene_counts</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="s1">&#39;results/total_count_table.tsv&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/total_count_table.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="n">script</span><span class="p">:</span>  <span class="c1"># Add script directive</span>
        <span class="s1">&#39;../scripts/count_table.py&#39;</span>  <span class="c1"># Add script location relative to rule file</span>
</code></pre></div></td></tr></table></div></p>
</details>
<details class="info">
<summary>Using the same Python script in and out of Snakemake</summary>
<p>To avoid code redundancy, it would be ideal to have a script that can be called by Snakemake but also work with standard Python (and be used outside Snakemake). The two main ways to do this are:</p>
<ul>
<li>Implement the script as a module/package and use this module in Snakemake, for example with a command line interface in <code>shell</code></li>
<li>Test whether the <code>snakemake</code> object exists in the script:<ul>
<li>If so, the script can process the Snakemake values<ul>
<li>When the script is launched by Snakemake, there is an object called <code>snakemake</code> that provides access to the same objects that are available in the <code>run</code> and <code>shell</code> directives (<code>input</code>, <code>output</code>, <code>params</code>, <code>wildcards</code>, <code>log</code>, <code>threads</code>, <code>resources</code>, config). For instance, you can use <code>snakemake.input[0]</code> to access the first input file of a rule, or <code>snakemake.input.input_name</code> to access a named input</li>
</ul>
</li>
<li>If not, the script can use other parameters, for example those coming from command line parsing</li>
</ul>
</li>
</ul>
</details>
<h5 id="providing-a-rule-specific-conda-environment">Providing a rule-specific conda environment</h5>
<p>Given the presence of a non-default package in the script, we need to find a solution to make it accessible inside the rule. The easiest way to do that is to create a rule-specific conda environment. In Snakemake, you can do this by providing an environment config file (in YAML format) to the rule with the <code>conda</code> directive.</p>
<p><strong>(Optional) Exercise:</strong> If you have time, you can create your own config file for the environment using the tip on &lsquo;Environment features&rsquo; below. If you need a reminder on how environment files look and work, you can check out slides 4-10 of the presentation (available <a href="#material">here</a>). Otherwise, you can directly skip to the answer.</p>
<details class="tip">
<summary>Environment features</summary>
<ul>
<li>Environment <code>name</code> is <code>py3.12</code></li>
<li>It uses two <code>channels</code>: <code>conda-forge</code> and <code>bioconda</code> (in that order)</li>
<li>It requires <code>python</code> with <strong>at least</strong> version <code>3.12</code></li>
<li>It requires <code>pandas</code> with version <code>2.2.3</code> <strong>exactly</strong></li>
<li>Like with scripts, config files should be stored in their own dedicated folder: <code>workflow/envs</code></li>
</ul>
</details>
<details class="success">
<summary>Answer</summary>
<p>The config file, created in <code>workflow/envs/py.yaml</code> should look like this:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Environment file to perform data processing with Python</span>
<span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">py3.12</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">python &gt;= 3.12</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pandas == 2.2.3</span>
</code></pre></div></td></tr></table></div></p>
</details>
<p><strong>Exercise:</strong> Add the conda environment to the rule.</p>
<details class="tip">
<summary>Environment file path is relative&hellip;</summary>
<p>&hellip; to the Snakefile calling it. If you followed the recommended workflow structure, modular snakefiles are placed in a <code>rules</code> subfolder (like <code>rules/analysis.smk</code>) and environment files are placed in a <code>envs</code> subfolder (like <code>envs/py.yaml</code>). You need to find a path between those two subfolders.</p>
</details>
<details class="success">
<summary>Answer</summary>
<p>We need to fill the last two missing elements with the directive name, <code>conda</code>, and its value, the script location:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="n">rule</span> <span class="n">count_table</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        This rule merges gene count tables of an assembly into one table.</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="nb">input</span><span class="p">:</span>
            <span class="n">get_gene_counts</span>
        <span class="n">output</span><span class="p">:</span>
            <span class="n">table</span> <span class="o">=</span> <span class="s1">&#39;results/total_count_table.tsv&#39;</span>
        <span class="n">log</span><span class="p">:</span>
            <span class="s1">&#39;logs/total_count_table.log&#39;</span>
        <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
        <span class="n">conda</span><span class="p">:</span>  <span class="c1"># Add conda directive</span>
            <span class="s1">&#39;../envs/py.yaml&#39;</span>  <span class="c1"># Add config file location relative to rule file</span>
        <span class="n">script</span><span class="p">:</span>
            <span class="s1">&#39;../scripts/count_table.py&#39;</span>
</code></pre></div></td></tr></table></div></p>
</details>
<p>Using conda environments improves reproducibility for many reasons, including version control and the fact that users do not need to manually manage software dependencies. The <strong>first workflow execution after adding Conda environments will take more time than usual</strong> because <code>snakemake</code> (through <code>conda</code>) has to <strong>download and install all the software</strong> in the working directory.</p>
<h4 id="adapting-the-snakefile-and-running-the-rule">Adapting the Snakefile and running the rule</h4>
<p>All that is left is running the rule to create the table.</p>
<p><strong>Exercise:</strong> Find the <code>snakemake</code> command you should run to create the desired output (which one is it?) and execute the workflow. Is there anything else to do beforehand?</p>
<details class="success">
<summary>Answer</summary>
<p>We cannot launch the workflow directly: first, we need to update rule <code>all</code> input to use the output of rule <code>count_table</code>. After this, your Snakefile should be:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">Main Snakefile of RNAseq analysis workflow. This workflow can clean and</span>
<span class="sd">map reads, and perform Differential Expression Analyses.</span>
<span class="sd">&#39;&#39;&#39;</span>

<span class="c1"># Config file path</span>
<span class="n">configfile</span><span class="p">:</span> <span class="s1">&#39;config/config.yaml&#39;</span>

<span class="c1"># Rules to execute workflow</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/read_mapping.smk&#39;</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/analyses.smk&#39;</span>

<span class="c1"># Master rule used to launch workflow</span>
<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Dummy rule to automatically generate required outputs.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
<span class="hll">        <span class="n">rules</span><span class="o">.</span><span class="n">count_table</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">table</span>  <span class="c1"># New input matching output of rule `count_table`</span>
</span></code></pre></div></td></tr></table></div></p>
<p>Finally, run the workflow with:
<div class="highlight"><pre><span></span><code>snakemake -c 4 -p --sdm conda
</code></pre></div>
Do not forget to add <code>--sdm conda</code>, otherwise Snakemake will not use the environment you provided.</p>
</details>
<p>You should see the rule <code>count_table</code> executed with 6 files as input. You can also check the log of rule <code>count_table</code> to see if the script worked as intended.</p>
<h3 id="creating-a-rule-to-detect-differentially-expressed-genes-deg">Creating a rule to detect Differentially Expressed Genes (DEG)</h3>
<p>The final rule that you will create in this course will use an R script to process the global read count table previously created and detect differentially expressed genes. As such, you will see several common elements between this rule and rule <code>count_table</code> (external scripts, dedicated environments, rule structure&hellip;) and the process to implement this rule will also be very similar. However, it will be easier as you won&rsquo;t need to use an input function.</p>
<h4 id="building-the-general-rule-structure-again">Building the general rule structure (again)</h4>
<p>We also wrote the common elements of the rule so that you can focus on the most interesting parts (the missing elements at the end):
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">rule</span> <span class="n">differential_expression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule detects DEGs and plots control graphs (PCA, heatmaps...).</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">count_table</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">table</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">deg</span> <span class="o">=</span> <span class="s1">&#39;results/deg_list.tsv&#39;</span><span class="p">,</span>
        <span class="n">pdf</span> <span class="o">=</span> <span class="s1">&#39;results/deg_plots.pdf&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/differential_expression.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
</code></pre></div></td></tr></table></div></p>
<p>Once again, we do not need to use <code>wildcards</code> in this rule, because all the files are precisely defined.</p>
<p><strong>Exercise:</strong> Given the rule structure above, update the <code>Snakefile</code> so that it creates the final output of the workflow.</p>
<details class="success">
<summary>Answer</summary>
<p>There is only one thing to update in the target rule input:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">Main Snakefile of RNAseq analysis workflow. This workflow can clean and</span>
<span class="sd">map reads, and perform Differential Expression Analyses.</span>
<span class="sd">&#39;&#39;&#39;</span>

<span class="c1"># Config file path</span>
<span class="n">configfile</span><span class="p">:</span> <span class="s1">&#39;config/config.yaml&#39;</span>

<span class="c1"># Rules to execute workflow</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/read_mapping.smk&#39;</span>
<span class="n">include</span><span class="p">:</span> <span class="s1">&#39;rules/analyses.smk&#39;</span>

<span class="c1"># Master rule used to launch workflow</span>
<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Dummy rule to automatically generate required outputs.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
<span class="hll">        <span class="n">rules</span><span class="o">.</span><span class="n">differential_expression</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">deg</span>  <span class="c1"># New input matching output of rule `differential_expression`</span>
</span></code></pre></div></td></tr></table></div>
Remember that we don&rsquo;t need to add both outputs of rule <code>differential_expression</code> as inputs of rule <code>all</code>, only one suffices.</p>
</details>
<p>As mentioned above, we don&rsquo;t need an input function because the input of rule <code>differential_expression</code> is easy to identify, so we&rsquo;ll directly focus on a way to run the R script.</p>
<h4 id="using-an-r-script-in-snakemake">Using an R script in Snakemake</h4>
<h5 id="downloading-the-script">Downloading the script</h5>
<p>The DE analyses will be performed thanks to a script called <code>DESeq2.R</code>. It was written in <a href="https://www.r-project.org/">R</a>, takes a read count table as input, and produces two outputs, a tab-separated table containing DEG (and statistical results) and a .pdf file containing control plots of the analysis. You can download it <a href="https://raw.githubusercontent.com/sib-swiss/containers-snakemake-training/main/docs/solutions/session4/workflow/scripts/DESeq2.R">here</a> or with the command:
<div class="highlight"><pre><span></span><code>wget<span class="w"> </span>https://raw.githubusercontent.com/sib-swiss/containers-snakemake-training/main/docs/solutions/session4/workflow/scripts/DESeq2.R
</code></pre></div></p>
<p><strong>Exercise:</strong> Download the script and place it the right folder.</p>
<details class="success">
<summary>Answer</summary>
<p>You can place this script in the same folder than the Python script, <code>workflow/scripts</code>. There is nothing in the official documentation about placing scripts from different languages in separate folders, but if you use a large number of scripts, it might be worth considering. You could also gather scripts by topic, similarly to <code>.smk</code> files.</p>
</details>
<h5 id="running-the-script">Running the script</h5>
<p>The next exercise won&rsquo;t be as guided as the other ones. This is done on purpose as you have seen everything you need to solve it!</p>
<p><strong>Exercise:</strong> Find a way to run the R script and fill the missing elements in the rule.</p>
<details class="tip">
<summary>What do you need to take into account?</summary>
<ul>
<li>The directive you need to run the script</li>
<li>The location/path of the script</li>
<li>Check whether the script need a special environment<ul>
<li>If so, and if you attended the <a href="https://www.sib.swiss/training/course/20250527_DOCK">SIB course on Containers</a>, remember a <em>certain Docker image</em> you created during the course</li>
<li>If not, use the following container: <code>docker://athiebaut/deseq2:v3</code></li>
</ul>
</li>
</ul>
</details>
<details class="success">
<summary>Answer</summary>
<p>Like with the Python script, there are two problems to solve to run the R script:</p>
<ol>
<li>
<p>Which directive to use?</p>
<p>There isn&rsquo;t much of a choice here&hellip; If you remember the presentation, there is only one way to run non-Python code in Snakemake: the <code>script</code> directive:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">rule</span> <span class="n">differential_expression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule detects DEGs and plots control graphs (PCA, heatmaps...).</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">count_table</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">table</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">deg</span> <span class="o">=</span> <span class="s1">&#39;results/deg_list.tsv&#39;</span><span class="p">,</span>
        <span class="n">pdf</span> <span class="o">=</span> <span class="s1">&#39;results/deg_plots.pdf&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/differential_expression.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="err">?</span><span class="p">:</span>
        <span class="err">?</span>
    <span class="n">script</span><span class="p">:</span>  <span class="c1"># Add script directive</span>
        <span class="s1">&#39;../scripts/DESeq2.R&#39;</span>  <span class="c1"># Add script location relative to rule file</span>
</code></pre></div></td></tr></table></div></p>
</li>
<li>
<p>Does it use external packages and need a specific environment?</p>
<p>If you look at the top of the script, you will see several (11 to be exact!) <code>library()</code> calls. Each of them imports an external package. All of these could be gathered in a conda environment, however when numerous libraries are involved, it is sometimes easier to use a container. During Session 3 of the Containers course (<a href="https://sib-swiss.github.io/containers-introduction-training/latest/course_material/day1/dockerfiles/#build-an-image-for-your-own-script">Working with Dockerfiles</a>), you built your own Docker image, called <code>deseq2</code>. This image actually contains everything required by the script:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">rule</span> <span class="n">differential_expression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This rule detects DEGs and plots control graphs (PCA, heatmaps...).</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">count_table</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">table</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">deg</span> <span class="o">=</span> <span class="s1">&#39;results/deg_list.tsv&#39;</span><span class="p">,</span>
        <span class="n">pdf</span> <span class="o">=</span> <span class="s1">&#39;results/deg_plots.pdf&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s1">&#39;logs/differential_expression.log&#39;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">container</span><span class="p">:</span>  <span class="c1"># Add container directive</span>
        <span class="s1">&#39;docker://athiebaut/deseq2:v3&#39;</span>  <span class="c1"># Try with your own image; if it doesn&#39;t work, use this one</span>
    <span class="n">script</span><span class="p">:</span>
        <span class="s1">&#39;../scripts/DESeq2.R&#39;</span>
</code></pre></div></td></tr></table></div>
</li>
</ol>
</details>
<details class="info">
<summary>Using the same R script in and out of Snakemake</summary>
<p>Inside the script, an S4 object named <code>snakemake</code>, analogous to the Python one, is available and allows access to Snakemake objects: <code>input</code>, <code>output</code>, <code>params</code>, <code>wildcards</code>, <code>log</code>, <code>threads</code>, <code>resources</code>, and <code>config</code>. Here, the syntax follows that of <a href="https://adv-r.hadley.nz/s4.html">S4 classes</a> with attributes that are R lists. For example, you can access the <strong>first</strong> input file with <code>snakemake@input[[1]]</code> (remember that in R, indexing starts at 1). Named objects can be accessed the same way, by providing the name instead of an index: <code>snakemake@input[["myfile"]]</code> to access the input called <code>myfile</code>.</p>
</details>
<p>Now, all that is left is running the workflow, check its outputs and visualise its DAG!</p>
<h4 id="running-the-workflow">Running the workflow</h4>
<p><strong>Exercise:</strong> Run the workflow. How many DEGs are detected during the analysis?</p>
<details class="success">
<summary>Answer</summary>
<p>You can run the workflow with:
<div class="highlight"><pre><span></span><code>snakemake -c 4 -p --sdm apptainer
</code></pre></div>
Do not forget to add <code>--sdm apptainer</code>, otherwise Snakemake will not pull the image and the script will be executed in the default environment (which will most likely lead to a crash).</p>
<p>During the run, you should see log messages about Snakemake managing the Docker image:
<div class="highlight"><pre><span></span><code>Pulling<span class="w"> </span>singularity<span class="w"> </span>image<span class="w"> </span>docker://athiebaut/deseq2:v3.
<span class="o">[</span>...<span class="o">]</span>
Activating<span class="w"> </span>singularity<span class="w"> </span>image<span class="w"> </span>/path/to/snakemake_rnaseq/.snakemake/singularity/8bfdbe93244feb95887ab5d33a705017.simg
</code></pre></div></p>
<p>To find how many genes are differentially expressed, check out the last output file, <code>results/deg_list.tsv</code>. 10 genes are differentially expressed in total: 4 up-regulated and 6 down-regulated.</p>
</details>
<details class="tip">
<summary>Containerisation of Conda-based workflows</summary>
<p>Snakemake can also automatically generate a Dockerfile that contains all required environments in a human readable way. If you want to see how a Snakemake-generated Dockerfile looks like, use:
<div class="highlight"><pre><span></span><code>snakemake -c 1 --containerize &gt; Dockerfile
</code></pre></div></p>
</details>
<p><strong>(Optional) Exercise:</strong> If you had to re-run the entire workflow from scratch, what command would you use?</p>
<details class="success">
<summary>Answer</summary>
<p>You can re-run the whole workflow with:
<div class="highlight"><pre><span></span><code>snakemake -c 4 -p -F --sdm conda apptainer
</code></pre></div></p>
<ul>
<li><code>-F</code> forces the execution of the entire workflow</li>
<li>Remember that you need <strong>both</strong> Conda and Docker-based environments for this run! You can combine <code>--sdm conda</code> and <code>--sdm apptainer</code> into a <strong>single command</strong> <code>--sdm conda apptainer</code>. Otherwise, you will lack some software and packages and the workflow will crash!</li>
</ul>
</details>
<p><strong>Exercise:</strong> Visualise the DAG of the entire workflow.</p>
<details class="success">
<summary>Answer</summary>
<p>You can get the DAG with:
<div class="highlight"><pre><span></span><code>snakemake -c 1 -p -F --dag | dot -T png &gt; images/total_dag.png
</code></pre></div></p>
<p>You should get the following DAG (open the picture in a new tab to zoom in):
<figure align="center">
  <img src="../../../assets/images/total_dag.png" width="100%"/>
</figure></p>
</details>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021-2025 SIB Swiss institute of Bioinformatics - <a href="#__consent">Change cookie settings</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss", "content.code.copy"], "search": "../../assets/javascripts/workers/search.07f07601.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.56dfad97.min.js"></script>
      
    
  </body>
</html>